README.md
[![Build Status](https://img.shields.io/travis/mikeu/mmsi.js)](https://travis-ci.org/mikeu/mmsi.js)
[![Code coverage](https://img.shields.io/codecov/c/github/mikeu/mmsi.js)](https://codecov.io/github/mikeu/mmsi.js)
[![Open issues](https://img.shields.io/github/issues/mikeu/mmsi.js)](https://github.com/mikeu/mmsi.js/issues)
[![NPM package](https://img.shields.io/npm/v/mmsi.js)](https://www.npmjs.com/package/mmsi.js)
[![License](https://img.shields.io/github/license/mikeu/mmsi.js)](https://github.com/mikeu/mmsi.js/blob/master/LICENSE)

# mmsi.js
A javascript library for parsing MMSI numbers (Maritime Mobile Service Identities) to validate format and extract
information such as country or other jurisdiction of registration.

## About
From [Wikipedia](https://en.wikipedia.org/wiki/Maritime_Mobile_Service_Identity):

> A Maritime Mobile Service Identity (MMSI) is a series of nine digits which are sent in digital form
> over a radio frequency channel in order to uniquely identify ship stations, ship earth stations,
> coast stations, coast earth stations, and group calls.

With mmsi.js you can determine whether a given number or string is in the format of a valid MMSI and,
if so, extract the
[Maritime Identification Digits](https://en.wikipedia.org/wiki/Maritime_identification_digits)
from it, along with the corresponding country name and code.

Note that this library makes no attempt to verify that a given MMSI is actually in active use, or registered to a real vessel.

## Installation
```bash
npm install mmsi.js
```

## Usage
```js
import { MMSI } from 'mmsi.js';

const mmsi = new MMSI(1234567890);
console.log(mmsi.isValid); // false -- too many digits
mmsi.identity = "ABCDEFGHI";
console.log(mmsi.isValid); // false -- not exclusively digits

mmsi.identity = 123;
console.log(mmsi.isValid); // true
console.log(mmsi.formatted); // "000000123"

mmsi.identity = "410123456";
console.log(mmsi.midCode); // "410"
const jurisdiction = mmsi.jurisdiction;
console.log(jurisdiction.code); // "BT"
console.log(jurisdiction.name); // "Bhutan"
console.log(jurisdiction.fullName); // "Bhutan (Kingdom of)"
```

## License
This project is licensed under the MIT License - see the
[LICENSE](https://github.com/mikeu/mmsi.js/blob/master/LICENSE)
file for details.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2.0] - 2020-02-10
### Changed
- Instead of having `identity` as a getter on the `MMSI` class for a "private" backing
field, that property is now used directly and the `formatted` getter has been added to
retrieve its zero-padded form, when required.

## [0.1.0] - 2020-02-09
### Changed
- Package now uses `mmsi.js` as its entrypoint, instead of the non-existent `index.js`.

## [0.0.1] - 2020-02-09
### Added
- Initial version.

tests/mmsi/mid.test.ts
import { MMSI } from "@mmsi";

test("It extracts MID codes from vessel MMSIs", () => {
  expect(new MMSI(231123456).midCode).toBe("231");
  expect(new MMSI(316123456).midCode).toBe("316");
  expect(new MMSI(445123456).midCode).toBe("445");
  expect(new MMSI(536123456).midCode).toBe("536");
  expect(new MMSI(664123456).midCode).toBe("664");
  expect(new MMSI(770123456).midCode).toBe("770");
});

test("It extracts MID codes from SAR aircraft MMSIs", () => {
  expect(new MMSI(111316123).midCode).toBe("316");
  expect(new MMSI(111512123).midCode).toBe("512");
  expect(new MMSI(111403567).midCode).toBe("403");
});

test("It extracts MID codes from MMSIs for craft associated with parent ships", () => {
  expect(new MMSI(983169876).midCode).toBe("316");
  expect(new MMSI(984459876).midCode).toBe("445");
  expect(new MMSI(986649876).midCode).toBe("664");
  expect(new MMSI(986089876).midCode).toBe("608");
});

test("It extracts MID codes from navigational aid MMSIs", () => {
  expect(new MMSI(993169876).midCode).toBe("316");
  expect(new MMSI(994459876).midCode).toBe("445");
  expect(new MMSI(996649876).midCode).toBe("664");
  expect(new MMSI(996089876).midCode).toBe("608");
});

test("It returns undefined when MMSIs do not include MID codes", () => {
  expect(new MMSI(970881234).midCode).toBeUndefined();
  expect(new MMSI(972881234).midCode).toBeUndefined();
  expect(new MMSI(974881234).midCode).toBeUndefined();
  expect(new MMSI(12345678).midCode).toBeUndefined();
  expect(new MMSI(123456789).midCode).toBeUndefined();
  expect(new MMSI(990000000).midCode).toBeUndefined();
  expect(new MMSI(111000000).midCode).toBeUndefined();
})

tests/mmsi.isValid.test.ts
import { MMSI } from "../src/index";

test("It returns true for valid numeric MMSIs", () => {
  expect(new MMSI(316123456).isValid).toBe(true);
  expect(new MMSI(512987654).isValid).toBe(true);
  expect(new MMSI(258881000).isValid).toBe(true);
  // MMSIs starting with zeros will be zero-padded.
  expect(new MMSI(31612345).isValid).toBe(true);
  expect(new MMSI(3161234).isValid).toBe(true);
});

test("It returns true for valid string MMSIs", () => {
  expect(new MMSI("316123456").isValid).toBe(true);
  expect(new MMSI("512987654").isValid).toBe(true);
  expect(new MMSI("258881000").isValid).toBe(true);
  expect(new MMSI("031612345").isValid).toBe(true);
  expect(new MMSI("003161234").isValid).toBe(true);
  expect(new MMSI("  3161234").isValid).toBe(true);
  expect(new MMSI("3161234").isValid).toBe(true);
})

test("It returns false for invalid MMSIs", () => {
  expect( new MMSI(3.14159).isValid).toBe(false);
  expect( new MMSI("MMSI").isValid).toBe(false);
  expect( new MMSI("99.999999").isValid).toBe(false);
});

tests/mmsi.midCode.test.ts
import { MMSI } from "../src/index";

test("It extracts MID codes from vessel MMSIs", () => {
  expect(new MMSI(231123456).midCode).toBe("231");
  expect(new MMSI(316123456).midCode).toBe("316");
  expect(new MMSI(445123456).midCode).toBe("445");
  expect(new MMSI(536123456).midCode).toBe("536");
  expect(new MMSI(664123456).midCode).toBe("664");
  expect(new MMSI(770123456).midCode).toBe("770");
});

test("It extracts MID codes from SAR aircraft MMSIs", () => {
  expect(new MMSI(111316123).midCode).toBe("316");
  expect(new MMSI(111512123).midCode).toBe("512");
  expect(new MMSI(111403567).midCode).toBe("403");
});

test("It extracts MID codes from MMSIs for craft associated with parent ships", () => {
  expect(new MMSI(983169876).midCode).toBe("316");
  expect(new MMSI(984459876).midCode).toBe("445");
  expect(new MMSI(986649876).midCode).toBe("664");
  expect(new MMSI(986089876).midCode).toBe("608");
});

test("It extracts MID codes from navigational aid MMSIs", () => {
  expect(new MMSI(993169876).midCode).toBe("316");
  expect(new MMSI(994459876).midCode).toBe("445");
  expect(new MMSI(996649876).midCode).toBe("664");
  expect(new MMSI(996089876).midCode).toBe("608");
});

test("It returns undefined when MMSIs do not include MID codes", () => {
  expect(new MMSI(970881234).midCode).toBeUndefined();
  expect(new MMSI(972881234).midCode).toBeUndefined();
  expect(new MMSI(974881234).midCode).toBeUndefined();
  expect(new MMSI(12345678).midCode).toBeUndefined();
  expect(new MMSI(123456789).midCode).toBeUndefined();
  expect(new MMSI(990000000).midCode).toBeUndefined();
  expect(new MMSI(111000000).midCode).toBeUndefined();
})

src/data.ts
import { Jurisdiction } from "@mmsi/jurisdiction";

/**
 * Get the country code corresponding to the given MID code, if it exists.
 *
 * In almost all cases the code is the ISO 3166 code for a country or other
 * jurisdiction, though not every MID code corresponds to an entity with an
 * ISO 3166 code.
 *
 * @param mid Maritime Identification Digits (MID code)
 * @returns Country code
 */
export const codeFromMID = (mid: string): string => {
  return midCodes[mid];
}

/**
 * Get the Jurisdiction corresponding to the given country code, if it exists.
 *
 * @param code ISO3166-based country code
 * @returns Jurisdiction for the code, or undefined.
 */
export const jurisdictionFromCode = (code: string): Jurisdiction => {
  return jurisdictions[code];
}

/**
 * Get the country object corresponding to the given MID code, if it exists.
 *
 * @param mid Maritime Identification Digits (MID code)
 * @returns Jurisdiction for the MID, or undefined.
 */
export const jurisdictionFromMID = (mid: string): Jurisdiction => {
  return jurisdictionFromCode(codeFromMID(mid));
}

const jurisdictions: { [code: string]: Jurisdiction } = {
  "AD": new Jurisdiction("AD", "Andorra", "Andorra (Principality of)"),
  "AE": new Jurisdiction("AE", "United Arab Emirates", "United Arab Emirates"),
  "AF": new Jurisdiction("AF", "Afghanistan"),
  "AG": new Jurisdiction("AG", "Antigua and Barbuda"),
  "AI": new Jurisdiction("AI", "Anguilla"),
  "AL": new Jurisdiction("AL", "Albania", "Albania (Republic of)"),
  "AM": new Jurisdiction("AM", "Armenia", "Armenia (Republic of)"),
  "AO": new Jurisdiction("AO", "Angola", "Angola (Republic of)"),
  "AR": new Jurisdiction("AR", "Argentina", "Argentine Republic"),
  "AS": new Jurisdiction("AS", "American Samoa"),
  "AT": new Jurisdiction("AT", "Austria"),
  "AU": new Jurisdiction("AU", "Australia"),
  "AW": new Jurisdiction("AW", "Aruba"),
  "AZ": new Jurisdiction("AZ", "Azerbaijan", "Azerbaijani Republic"),
  "BA": new Jurisdiction("BA", "Bosnia and Herzegovina"),
  "BB": new Jurisdiction("BB", "Barbados"),
  "BD": new Jurisdiction("BD", "Bangladesh", "Bangladesh (People's Republic of)"),
  "BE": new Jurisdiction("BE", "Belgium"),
  "BF": new Jurisdiction("BF", "Burkina Faso"),
  "BG": new Jurisdiction("BG", "Bulgaria", "Bulgaria (Republic of)"),
  "BH": new Jurisdiction("BH", "Bahrain", "Bahrain (Kingdom of)"),
  "BI": new Jurisdiction("BI", "Burundi", "Burundi (Republic of)"),
  "BJ": new Jurisdiction("BJ", "Benin", "Benin (Republic of)"),
  "BM": new Jurisdiction("BM", "Bermuda"),
  "BN": new Jurisdiction("BN", "Brunei", "Brunei Darussalam"),
  "BO": new Jurisdiction("BO", "Bolivia", "Bolivia (Republic of)"),
  "BQ,CW,SX": new Jurisdiction(
    "BQ,CW,SX",
    "Caribbean Netherlands; Curaçao; Sint Maarten",
    "Bonaire, Sint Eustatius and Saba; Curaçao; Sint Maarten - Netherlands (Kingdom of the)"
  ),
  "BR": new Jurisdiction("BR", "Brazil", "Brazil (Federative Republic of)"),
  "BS": new Jurisdiction("BS", "Bahamas", "Bahamas (Commonwealth of the)"),
  "BT": new Jurisdiction("BT", "Bhutan", "Bhutan (Kingdom of)"),
  "BW": new Jurisdiction("BW", "Botswana", "Botswana (Republic of)"),
  "BY": new Jurisdiction("BY", "Belarus", "Belarus (Republic of)"),
  "BZ": new Jurisdiction("BZ", "Belize"),
  "CA": new Jurisdiction("CA", "Canada"),
  "CC": new Jurisdiction("CC", "Cocos", "Cocos (Keeling) Islands"),
  "CD": new Jurisdiction("CD", "Democratic Republic of the Congo"),
  "CF": new Jurisdiction("CF", "Central African Republic"),
  "CG": new Jurisdiction("CG", "Congo", "Congo (Republic of the)"),
  "CH": new Jurisdiction("CH", "Switzerland", "Switzerland (Confederation of)"),
  "CI": new Jurisdiction("CI", "Côte d'Ivoire", "Côte d'Ivoire (Republic of)"),
  "CK": new Jurisdiction("CK", "Cook Islands"),
  "CL": new Jurisdiction("CL", "Chile"),
  "CM": new Jurisdiction("CM", "Cameroon", "Cameroon (Republic of)"),
  "CN": new Jurisdiction("CN", "China", "China (People's Republic of)"),
  "CO": new Jurisdiction("CO", "Colombia", "Colombia (Republic of)"),
  "CR": new Jurisdiction("CR", "Costa Rica"),
  "CU": new Jurisdiction("CU", "Cuba"),
  "CV": new Jurisdiction("CV", "Cape Verde", "Cape Verde (Republic of)"),
  "CX": new Jurisdiction("CX", "Christmas Island", "Christmas Island (Indian Ocean)"),
  "CY": new Jurisdiction("CY", "Cyprus", "Cyprus (Republic of)"),
  "CZ": new Jurisdiction("CZ", "Czech Republic", "Czech Republic"),
  "DE": new Jurisdiction("DE", "Germany", "Germany (Federal Republic of)"),
  "DJ": new Jurisdiction("DJ", "Djibouti", "Djibouti (Republic of)"),
  "DK": new Jurisdiction("DK", "Denmark"),
  "DM": new Jurisdiction("DM", "Dominica", "Dominica (Commonwealth of)"),
  "DO": new Jurisdiction("DO", "Dominican Republic", "Dominican Republic"),
  "DZ": new Jurisdiction("DZ", "Algeria", "Algeria (People's Democratic Republic of)"),
  "EC": new Jurisdiction("EC", "Ecuador"),
  "EE": new Jurisdiction("EE", "Estonia", "Estonia (Republic of)"),
  "EG": new Jurisdiction("EG", "Egypt", "Egypt (Arab Republic of)"),
  "ER": new Jurisdiction("ER", "Eritrea"),
  "ES": new Jurisdiction("ES", "Spain"),
  "ET": new Jurisdiction("ET", "Ethiopia", "Ethiopia (Federal Democratic Republic of)"),
  "FI": new Jurisdiction("FI", "Finland"),
  "FJ": new Jurisdiction("FJ", "Fiji", "Fiji (Republic of)"),
  "FK": new Jurisdiction("FK", "Falkland Islands", "Falkland Islands (Malvinas)"),
  "FM": new Jurisdiction("FM", "Micronesia", "Micronesia (Federated States of)"),
  "FO": new Jurisdiction("FO", "Faroe Islands"),
  "FR": new Jurisdiction("FR", "France"),
  "GA": new Jurisdiction("GA", "Gabonese Republic"),
  "GB": new Jurisdiction("GB", "United Kingdom", "United Kingdom of Great Britain and Northern Ireland"),
  "GD": new Jurisdiction("GD", "Grenada"),
  "GE": new Jurisdiction("GE", "Georgia"),
  "GF": new Jurisdiction("GF", "Guiana", "Guiana (French Department of)"),
  "GH": new Jurisdiction("GH", "Ghana"),
  "GI": new Jurisdiction("GI", "Gibraltar"),
  "GL": new Jurisdiction("GL", "Greenland"),
  "GM": new Jurisdiction("GM", "Gambia", "Gambia (Republic of the)"),
  "GN": new Jurisdiction("GN", "Guinea", "Guinea (Republic of)"),
  "GP": new Jurisdiction("GP", "Guadeloupe", "Guadeloupe (French Department of)"),
  "GQ": new Jurisdiction("GQ", "Equatorial Guinea", "Equatorial Guinea (Republic of)"),
  "GR": new Jurisdiction("GR", "Greece"),
  "GT": new Jurisdiction("GT", "Guatemala", "Guatemala (Republic of)"),
  "GW": new Jurisdiction("GW", "Guinea-Bissau", "Guinea-Bissau (Republic of)"),
  "GY": new Jurisdiction("GY", "Guyana"),
  "HK": new Jurisdiction("HK", "Hong Kong", "Hong Kong (Special Administrative Region of China)"),
  "HN": new Jurisdiction("HN", "Honduras", "Honduras (Republic of)"),
  "HR": new Jurisdiction("HR", "Croatia", "Croatia (Republic of)"),
  "HT": new Jurisdiction("HT", "Haiti", "Haiti (Republic of)"),
  "HU": new Jurisdiction("HU", "Hungary", "Hungary (Republic of)"),
  "ID": new Jurisdiction("ID", "Indonesia", "Indonesia (Republic of)"),
  "IE": new Jurisdiction("IE", "Ireland"),
  "IL": new Jurisdiction("IL", "Israel", "Israel (State of)"),
  "IN": new Jurisdiction("IN", "India", "India (Republic of)"),
  "IQ": new Jurisdiction("IQ", "Iraq", "Iraq (Republic of)"),
  "IR": new Jurisdiction("IR", "Iran", "Iran (Islamic Republic of)"),
  "IS": new Jurisdiction("IS", "Iceland"),
  "IT": new Jurisdiction("IT", "Italy"),
  "JM": new Jurisdiction("JM", "Jamaica"),
  "JO": new Jurisdiction("JO", "Jordan", "Jordan (Hashemite Kingdom of)"),
  "JP": new Jurisdiction("JP", "Japan"),
  "KE": new Jurisdiction("KE", "Kenya", "Kenya (Republic of)"),
  "KG": new Jurisdiction("KG", "Kyrgyzstan"),
  "KH": new Jurisdiction("KH", "Cambodia", "Cambodia (Kingdom of)"),
  "KI": new Jurisdiction("KI", "Kiribati", "Kiribati (Republic of)"),
  "KM": new Jurisdiction("KM", "Comoros", "Comoros (Union of the)"),
  "KN": new Jurisdiction("KN", "Saint Kitts and Nevis"),
  "KP": new Jurisdiction("KP", "North Korea"),
  "KR": new Jurisdiction("KR", "Korea", "Korea (Republic of)"),
  "KW": new Jurisdiction("KW", "Kuwait", "Kuwait (State of)"),
  "KY": new Jurisdiction("KY", "Cayman Islands"),
  "KZ": new Jurisdiction("KZ", "Kazakhstan", "Kazakhstan (Republic of)"),
  "LA": new Jurisdiction("LA", "Laos", "Lao People's Democratic Republic (Laos)"),
  "LB": new Jurisdiction("LB", "Lebanon"),
  "LC": new Jurisdiction("LC", "Saint Lucia"),
  "LI": new Jurisdiction("LI", "Liechtenstein", "Liechtenstein (Principality of)"),
  "LK": new Jurisdiction("LK", "Sri Lanka", "Sri Lanka (Democratic Socialist Republic of)"),
  "LR": new Jurisdiction("LR", "Liberia", "Liberia (Republic of)"),
  "LS": new Jurisdiction("LS", "Lesotho", "Lesotho (Kingdom of)"),
  "LT": new Jurisdiction("LT", "Lithuania", "Lithuania (Republic of)"),
  "LU": new Jurisdiction("LU", "Luxembourg", "Luxembourg"),
  "LV": new Jurisdiction("LV", "Latvia", "Latvia (Republic of)"),
  "LY": new Jurisdiction("LY", "Libya"),
  "MA": new Jurisdiction("MA", "Morocco", "Morocco (Kingdom of)"),
  "MC": new Jurisdiction("MC", "Monaco", "Monaco (Principality of)"),
  "MD": new Jurisdiction("MD", "Moldova", "Moldova (Republic of)"),
  "ME": new Jurisdiction("ME", "Montenegro", "Montenegro (Republic of)"),
  "MG": new Jurisdiction("MG", "Madagascar", "Madagascar (Republic of)"),
  "MH": new Jurisdiction("MH", "Marshall Islands", "Marshall Islands (Republic of the)"),
  "MK": new Jurisdiction("MK", "Macedonia"),
  "ML": new Jurisdiction("ML", "Mali", "Mali (Republic of)"),
  "MM": new Jurisdiction("MM", "Myanmar", "Myanmar (Union of)"),
  "MN": new Jurisdiction("MN", "Mongolia"),
  "MO": new Jurisdiction("MO", "Macao"),
  "MP": new Jurisdiction("MP", "Northern Mariana Islands", "Northern Mariana Islands (Commonwealth of the)"),
  "MQ": new Jurisdiction("MQ", "Martinique", "Martinique (French Department of)"),
  "MR": new Jurisdiction("MR", "Mauritania", "Mauritania (Islamic Republic of)"),
  "MS": new Jurisdiction("MS", "Montserrat"),
  "MT": new Jurisdiction("MT", "Malta"),
  "MU": new Jurisdiction("MU", "Mauritius", "Mauritius (Republic of)"),
  "MV": new Jurisdiction("MV", "Maldives", "Maldives (Republic of)"),
  "MW": new Jurisdiction("MW", "Malawi"),
  "MX": new Jurisdiction("MX", "Mexico"),
  "MY": new Jurisdiction("MY", "Malaysia"),
  "MZ": new Jurisdiction("MZ", "Mozambique", "Mozambique (Republic of)"),
  "NA": new Jurisdiction("NA", "Namibia", "Namibia (Republic of)"),
  "NC": new Jurisdiction("NC", "New Caledonia", "New Caledonia"),
  "NE": new Jurisdiction("NE", "Niger", "Niger (Republic of the)"),
  "NG": new Jurisdiction("NG", "Nigeria", "Nigeria (Federal Republic of)"),
  "NI": new Jurisdiction("NI", "Nicaragua"),
  "NL": new Jurisdiction("NL", "Netherlands", "Netherlands (Kingdom of the)"),
  "NO": new Jurisdiction("NO", "Norway"),
  "NP": new Jurisdiction("NP", "Nepal"),
  "NR": new Jurisdiction("NR", "Nauru", "Nauru (Republic of)"),
  "NU": new Jurisdiction("NU", "Niue"),
  "NZ": new Jurisdiction("NZ", "New Zealand"),
  "OM": new Jurisdiction("OM", "Oman", "Oman (Sultanate of)"),
  "PA": new Jurisdiction("PA", "Panama", "Panama (Republic of)"),
  "PE": new Jurisdiction("PE", "Peru"),
  "PF": new Jurisdiction("PF", "French Polynesia"),
  "PG": new Jurisdiction("PG", "Papua New Guinea"),
  "PH": new Jurisdiction("PH", "Philippines", "Philippines (Republic of the)"),
  "PK": new Jurisdiction("PK", "Pakistan", "Pakistan (Islamic Republic of)"),
  "PL": new Jurisdiction("PL", "Poland", "Poland (Republic of)"),
  "PM": new Jurisdiction("PM", "Saint Pierre and Miquelon", "Saint Pierre and Miquelon (Territorial Collectivity of)"),
  "PN": new Jurisdiction("PN", "Pitcairn Island"),
  "PR": new Jurisdiction("PR", "Puerto Rico"),
  "PS": new Jurisdiction("PS", "Palestinian Authority", "Palestinian Authority (based on Resolution 99 of PP-98)"),
  "PT": new Jurisdiction("PT", "Portugal"),
  "PT-20": new Jurisdiction("PT-20", "Azores", "Azores (Portuguese isles of)"),
  "PT-30": new Jurisdiction("PT-30", "Madeira", "Madeira (Portuguese isles of)"),
  "PW": new Jurisdiction("PW", "Palau", "Palau (Republic of)"),
  "PY": new Jurisdiction("PY", "Paraguay", "Paraguay (Republic of)"),
  "QA": new Jurisdiction("QA", "Qatar", "Qatar (State of)"),
  "RE": new Jurisdiction("RE", "Réunion", "Réunion (French Department of)"),
  "RO": new Jurisdiction("RO", "Romania"),
  "RS": new Jurisdiction("RS", "Serbia"),
  "RU": new Jurisdiction("RU", "Russia", "Russian Federation"),
  "RW": new Jurisdiction("RW", "Rwanda", "Rwandese Republic"),
  "SA": new Jurisdiction("SA", "Saudi Arabia", "Saudi Arabia (Kingdom of)"),
  "SB": new Jurisdiction("SB", "Solomon Islands"),
  "SC": new Jurisdiction("SC", "Seychelles", "Seychelles (Republic of)"),
  "SD": new Jurisdiction("SD", "Sudan", "Sudan (Republic of the)"),
  "SE": new Jurisdiction("SE", "Sweden"),
  "SG": new Jurisdiction("SG", "Singapore", "Singapore (Republic of)"),
  "SH-AC": new Jurisdiction("SH-AC", "Ascension Island"),
  "SH-HL": new Jurisdiction("SH-HL", "Saint Helena"),
  "SI": new Jurisdiction("SI", "Slovenia", "Slovenia (Republic of)"),
  "SK": new Jurisdiction("SK", "Slovakia"),
  "SL": new Jurisdiction("SL", "Sierra Leone"),
  "SM": new Jurisdiction("SM", "San Marino", "San Marino (Republic of)"),
  "SN": new Jurisdiction("SN", "Senegal", "Senegal (Republic of)"),
  "SO": new Jurisdiction("SO", "Somalia"),
  "SR": new Jurisdiction("SR", "Suriname", "Suriname (Republic of)"),
  "SS": new Jurisdiction("SS", "South Sudan", "South Sudan (Republic of the)"),
  "ST": new Jurisdiction("ST", "São Tomé and Príncipe", "São Tomé and Príncipe (Democratic Republic of)"),
  "SV": new Jurisdiction("SV", "El Salvador", "El Salvador (Republic of)"),
  "SY": new Jurisdiction("SY", "Syria", "Syrian Arab Republic"),
  "SZ": new Jurisdiction("SZ", "Swaziland", "Swaziland (Kingdom of)"),
  "TC": new Jurisdiction("TC", "Turks and Caicos Islands"),
  "TD": new Jurisdiction("TD", "Chad", "Chad (Republic of)"),
  "TF": new Jurisdiction("TF", "Adelie Land", "Adelie Land (French Southern Territories)"),
  "TF-CA": new Jurisdiction("TF-CA", "Crozet Archipelago"),
  "TF-KI": new Jurisdiction("TF-KI", "Kerguelen Islands"),
  "TF-SP": new Jurisdiction("TF-SP", "Saint Paul and Amsterdam Islands"),
  "TG": new Jurisdiction("TG", "Togo", "Togolese Republic"),
  "TH": new Jurisdiction("TH", "Thailand"),
  "TJ": new Jurisdiction("TJ", "Tajikistan", "Tajikistan (Republic of)"),
  "TM": new Jurisdiction("TM", "Turkmenistan"),
  "TN": new Jurisdiction("TN", "Tunisia"),
  "TO": new Jurisdiction("TO", "Tonga", "Tonga (Kingdom of)"),
  "TR": new Jurisdiction("TR", "Turkey"),
  "TT": new Jurisdiction("TT", "Trinidad and Tobago"),
  "TV": new Jurisdiction("TV", "Tuvalu"),
  "TW": new Jurisdiction("TW", "Taiwan", "Taiwan (Republic of China)"),
  "TZ": new Jurisdiction("TZ", "Tanzania", "Tanzania (United Republic of)"),
  "UA": new Jurisdiction("UA", "Ukraine"),
  "UG": new Jurisdiction("UG", "Uganda", "Uganda (Republic of)"),
  "US": new Jurisdiction("US", "United States of America"),
  "US-AK": new Jurisdiction("US-AK", "Alaska", "Alaska (State of)"),
  "UY": new Jurisdiction("UY", "Uruguay", "Uruguay (Eastern Republic of)"),
  "UZ": new Jurisdiction("UZ", "Uzbekistan"),
  "VA": new Jurisdiction("VA", "Vatican City", "Vatican City State"),
  "VC": new Jurisdiction("VC", "Saint Vincent and the Grenadines"),
  "VE": new Jurisdiction("VE", "Venezuela", "Venezuela (Bolivarian Republic of)"),
  "VG": new Jurisdiction("VG", "British Virgin Islands"),
  "VI": new Jurisdiction("VI", "US Virgin Islands", "United States Virgin Islands"),
  "VN": new Jurisdiction("VN", "Vietnam", "Vietnam (Socialist Republic of)"),
  "VU": new Jurisdiction("VU", "Vanuatu", "Vanuatu (Republic of)"),
  "WF": new Jurisdiction("WF", "Wallis and Futuna Islands"),
  "WS": new Jurisdiction("WS", "Samoa", "Samoa (Independent State of)"),
  "YE": new Jurisdiction("YE", "Yemen", "Yemen (Republic of)"),
  "ZA": new Jurisdiction("ZA", "South Africa", "South Africa (Republic of)"),
  "ZM": new Jurisdiction("ZM", "Zambia", "Zambia (Republic of)"),
  "ZW": new Jurisdiction("ZW", "Zimbabwe", "Zimbabwe (Republic of)"),
}

const midCodes: { [mid: string]: string } = {
  "201": "AL",
  "202": "AD",
  "203": "AT",
  "204": "PT-20",
  "205": "BE",
  "206": "BY",
  "207": "BG",
  "208": "VA",
  "209": "CY",
  "210": "CY",
  "212": "CY",
  "211": "DE",
  "218": "DE",
  "213": "GE",
  "214": "MD",
  "215": "MT",
  "229": "MT",
  "248": "MT",
  "249": "MT",
  "256": "MT",
  "216": "AM",
  "219": "DK",
  "220": "DK",
  "224": "ES",
  "225": "ES",
  "226": "FR",
  "227": "FR",
  "228": "FR",
  "230": "FI",
  "231": "FO",
  "232": "GB",
  "233": "GB",
  "234": "GB",
  "235": "GB",
  "236": "GI",
  "237": "GR",
  "239": "GR",
  "240": "GR",
  "241": "GR",
  "238": "HR",
  "242": "MA",
  "243": "HU",
  "244": "NL",
  "245": "NL",
  "246": "NL",
  "247": "IT",
  "250": "IE",
  "251": "IS",
  "252": "LI",
  "253": "LU",
  "254": "MC",
  "255": "PT-30",
  "257": "NO",
  "258": "NO",
  "259": "NO",
  "261": "PL",
  "262": "ME",
  "263": "PT",
  "264": "RO",
  "265": "SE",
  "266": "SE",
  "267": "SK",
  "268": "SM",
  "269": "CH",
  "270": "CZ",
  "271": "TR",
  "272": "UA",
  "273": "RU",
  "274": "MK",
  "275": "LV",
  "276": "EE",
  "277": "LT",
  "278": "SI",
  "279": "RS",
  "301": "AI",
  "303": "US-AK",
  "304": "AG",
  "305": "AG",
  "306": "BQ,CW,SX",
  "307": "AW",
  "308": "BS",
  "309": "BS",
  "311": "BS",
  "310": "BM",
  "312": "BZ",
  "314": "BB",
  "316": "CA",
  "319": "KY",
  "321": "CR",
  "323": "CU",
  "325": "DM",
  "327": "DO",
  "329": "GP",
  "330": "GD",
  "331": "GL",
  "332": "GT",
  "334": "HN",
  "336": "HT",
  "338": "US",
  "366": "US",
  "367": "US",
  "368": "US",
  "369": "US",
  "339": "JM",
  "341": "KN",
  "343": "LC",
  "345": "MX",
  "347": "MQ",
  "348": "MS",
  "350": "NI",
  "351": "PA",
  "352": "PA",
  "353": "PA",
  "354": "PA",
  "355": "PA",
  "356": "PA",
  "357": "PA",
  "370": "PA",
  "371": "PA",
  "372": "PA",
  "373": "PA",
  "374": "PA",
  "358": "PR",
  "359": "SV",
  "361": "PM",
  "362": "TT",
  "364": "TC",
  "375": "VC",
  "376": "VC",
  "377": "VC",
  "378": "VG",
  "379": "VI",
  "401": "AF",
  "403": "SA",
  "405": "BD",
  "408": "BH",
  "410": "BT",
  "412": "CN",
  "413": "CN",
  "414": "CN",
  "416": "TW",
  "417": "LK",
  "419": "IN",
  "422": "IR",
  "423": "AZ",
  "425": "IQ",
  "428": "IL",
  "431": "JP",
  "432": "JP",
  "434": "TM",
  "436": "KZ",
  "437": "UZ",
  "438": "JO",
  "440": "KR",
  "441": "KR",
  "443": "PS",
  "445": "KP",
  "447": "KW",
  "450": "LB",
  "451": "KG",
  "453": "MO",
  "455": "MV",
  "457": "MN",
  "459": "NP",
  "461": "OM",
  "463": "PK",
  "466": "QA",
  "468": "SY",
  "470": "AE",
  "471": "AE",
  "472": "TJ",
  "473": "YE",
  "475": "YE",
  "477": "HK",
  "478": "BA",
  "501": "TF",
  "503": "AU",
  "506": "MM",
  "508": "BN",
  "510": "FM",
  "511": "PW",
  "512": "NZ",
  "514": "KH",
  "515": "KH",
  "516": "CX",
  "518": "CK",
  "520": "FJ",
  "523": "CC",
  "525": "ID",
  "529": "KI",
  "531": "LA",
  "533": "MY",
  "536": "MP",
  "538": "MH",
  "540": "NC",
  "542": "NU",
  "544": "NR",
  "546": "PF",
  "548": "PH",
  "553": "PG",
  "555": "PN",
  "557": "SB",
  "559": "AS",
  "561": "WS",
  "563": "SG",
  "564": "SG",
  "565": "SG",
  "566": "SG",
  "567": "TH",
  "570": "TO",
  "572": "TV",
  "574": "VN",
  "576": "VU",
  "577": "VU",
  "578": "WF",
  "601": "ZA",
  "603": "AO",
  "605": "DZ",
  "607": "TF-SP",
  "608": "SH-AC",
  "609": "BI",
  "610": "BJ",
  "611": "BW",
  "612": "CF",
  "613": "CM",
  "615": "CG",
  "616": "KM",
  "620": "KM",
  "617": "CV",
  "618": "TF-CA",
  "619": "CI",
  "621": "DJ",
  "622": "EG",
  "624": "ET",
  "625": "ER",
  "626": "GA",
  "627": "GH",
  "629": "GM",
  "630": "GW",
  "631": "GQ",
  "632": "GN",
  "633": "BF",
  "634": "KE",
  "635": "TF-KI",
  "636": "LR",
  "637": "LR",
  "638": "SS",
  "642": "LY",
  "644": "LS",
  "645": "MU",
  "647": "MG",
  "649": "ML",
  "650": "MZ",
  "654": "MR",
  "655": "MW",
  "656": "NE",
  "657": "NG",
  "659": "NA",
  "660": "RE",
  "661": "RW",
  "662": "SD",
  "663": "SN",
  "664": "SC",
  "665": "SH-HL",
  "666": "SO",
  "667": "SL",
  "668": "ST",
  "669": "SZ",
  "670": "TD",
  "671": "TG",
  "672": "TN",
  "674": "TZ",
  "677": "TZ",
  "675": "UG",
  "676": "CD",
  "678": "ZM",
  "679": "ZW",
  "701": "AR",
  "710": "BR",
  "720": "BO",
  "725": "CL",
  "730": "CO",
  "735": "EC",
  "740": "FK",
  "745": "GF",
  "750": "GY",
  "755": "PY",
  "760": "PE",
  "765": "SR",
  "770": "UY",
  "775": "VE",
};